docs/Brew-Livecheck.md

```diff
@@ -0,0 +1,115 @@
+
+* For technical reasons, **the `regex` call in the `livecheck` block should always use parentheses** (e.g. `regex(/example/)`).
+
+* **Regex literals should follow the established Homebrew style**, where the `/.../` syntax is the default and the `%r{...}` syntax is used when a forward slash (`/`) is present.
+
+* **Regexes should adequately represent their intention.** This requires an understanding of basic regex syntax, so we avoid issues like using `.` (match any character) instead of `\.` when we only want to match a period. We also try to be careful about our use of generic catch-alls like  `.*` or `.+`, as it's often better to use something non-greedy and contextually appropriate. For example, if we wanted to match a variety of characters while trying to stay within the bounds of an HTML attribute, we could use something like `[^"' >]*?`.
+
+* **Try not to be too specific in some parts of the regex.** For example, if a file name uses a hyphen (`-`) between the software name and version (e.g. `example-1.2.3.tar.gz`), we may want to use something like `example[._-]v?(\d+(?:\.\d+)+)\.t` instead. This would allow the regex to continue matching if the upstream file name format changes to `example.1.2.3.tar.gz` or `example_1.2.3.tar.gz`.
+
+* **Regexes should be case insensitive unless case sensitivity is explicitly required for matching to work properly**. Case insensitivity is enabled by adding the `i` flag at the end of the regex literal (e.g. `/.../i` or `%r{...}i`). This helps to improve reliability and reduce maintenance, as a case-insensitive regex doesn't need to be manually updated if there are any upstream changes in letter case.
+
+* **Regexes should only use a capturing group around the part of the matched text that corresponds to the version**. For example, in `/href=.*?example-v?(\d+(?:\.\d+)+)(?:-src)?\.t/i`, we're only using a capturing group around the version part (matching a version like `1.2`, `1.2.3`, etc.) and we're using non-capturing groups elsewhere (e.g. `(?:-src)?`). This allows livecheck to rely on the first capture group being the version string.
+
+* **Regexes should only match stable versions**. Regexes should be written to avoid prerelease versions like `1.2-alpha1`, `1.2-beta1`, `1.2-rc1`, etc.
+
+* **Restrict matching to `href` attributes when targeting file names in an HTML page (or `url` attributes in an RSS feed)**. Using `href=.*?` (or `url=.*?`) at the start of the regex will take care of any opening delimiter for the attribute (`"`, `'`, or nothing) as well as any leading part of the URL. This helps to keep the regex from being overly specific, reducing the need for maintenance in the future. A regex like `href=.*?example-...` is often fine but sometimes it's necessary to have something explicit before the file name to limit matching to only what's appropriate (e.g. `href=.*?/example-...` or `href=["']?example-...`). Similarly, `["' >]` can be used to target the end of the attribute, when needed.
```

Comment:
I'm not sure what homebrew wants from me here. Should I use href? Should I not? What do I need to use it with?

Can we work with some examples here?

Reply:
Having read this I have no idea when I should or shouldn't use this.

Reply:
Basically, if we're matching a file name on an HTML page, restricting matching to `href` attribute URLs will reduce the scope and help to avoid unwanted matches. I reworked this particular guideline to the following:

* **Anchor the start/end of the regex, to restrict the scope**. For example, on HTML pages we often match file names or version directories in `href` attribute URLs (e.g. `/href=.*?example[._-]v?(\d+(?:\.\d+)+)\.zip/i`). The general idea is that limiting scope will help exclude unwanted matches.

Some of the previous information is handled by related examples in the new "Example `livecheck` blocks" section, which hopefully does a better job of providing context. Let me know if these areas are still unclear and I'll try to clarify further.

Reply:
@samford This is tangential, but as I’m seeing `(\d+(?:\.\d+)*)` a lot, why is such a complex regex needed? Specifically, which case does `(\d+(?:\.\d+)*)` correctly catch that `([\d\.]+)` wouldn’t?

Reply:
The reason we use `v?(\d+(?:\.\d+)+)` (note the `+` instead of `*`) is because it will ensure we're only matching something like `1.2`/`v1.2`, `1.2.3`/`v1.2.3`, etc. When I started working on improving livecheck over a year ago, `[0-9.]+` was used to match version text and the looseness was sometimes breaking checks due to it matching unwanted text and returning it as the newest version.

We created the more explicit regex in response to those issues and have standardized on using it when we want to match a version like `1.2`/`v1.2.3`, etc. People see the relative complexity of this regex and suggest something simpler like `[\d.]+` but we have experience with that not being sufficient.

The only problem we've had with the `+` regex is that it sometimes needs to be loosened to `v?(\d+(?:\.\d+)*)` (i.e., `*` instead of `+`) but this doesn't happen very often. For example, we use the `*` regex if a project predictably publishes the first release for a major version as `2` instead of `2.0`/`2.0.0` (e.g., `openjdk`). We also sometimes use the `*` regex for date-based versions that don't use delimiters (e.g., YYYYMMDD) but we're often more explicit with the regex when a date-based version does use a delimiter (e.g., `(\d{2,4}-\d{2}-\d{2})`).

Despite the looseness of the `*` regex, it's still better than `[0-9.]+` as it won't match text that isn't a complete version like `.`, `.1`, etc. It's explicit enough that it will only match text like `1`, `123`, `1.0`, etc. and this helps to avoid unexpected situations.

The `+` version is what we use by default (when we won't encounter versions like `1`, `123`, etc.) as the looser `*` version can match something it shouldn't if the text is only composed of digits. For example, a project may use versions like `1.2.3` for stable releases but also publish snapshots using date-based versions like `20210113`. If they both use the same format and are found in the same place, then the looser regex would match both and the date-based version would be erroneously returned as newest. This is something we've encountered a number of times, so this is why we use the `+` regex instead of `*` by default.

To put it into perspective, there are only 45 instances of the `*` regex in homebrew/core, compared to 1000+ instances of the `+` regex.

---

However, I imagine you're primarily seeing `(\d+(?:\.\d+)*)` in homebrew/cask, where there are 330 instances of it (and growing fast). homebrew/cask only has 24 instances of the `+` regex (all created by me, it seems).

Support for casks in livecheck was only recently added and, unfortunately, homebrew/cask has standardized on the wrong regex. I explained this concept in PR review (i.e., use the `+` regex unless `*` is necessary) but it was ignored, unfortunately. I also don't have write access in homebrew/cask, so I don't regularly review `livecheck` block PRs over there (as I can't merge them) like I do in homebrew/core.

Rather than trying to keep up with the pace that `livecheck` blocks are being added to casks, I'm currently trying to focus on adding more RuboCops that will automatically enforce the most of the guidelines. We have some livecheck RuboCops for formulae but they don't apply to casks yet, so I need to figure out how to make them work for both.

---

Regarding the complexity of this regex compared to `[0-9.]+` (or `[\d.]+`), these regexes will be replaced by constants, so the complexity will be abstracted away. At that point, we'll get the benefit of the additional specificity (over something loose like `[0-9.]+`) without asking people to engage with the additional complexity. [I imagine this may also help to improve the current situation with cask regexes not always following guidelines.]

I'm aiming to create a PR that adds these constants soon (the work's pretty much done) but I'm working to finish/merge some other PRs first.

I'm planning to eventually spend time bringing the `livecheck` blocks in homebrew/cask up to standard but I don't plan to work on this until sometime after I've implemented regex constants (as that will require modifying regexes across the entire tap anyway).